DESIGN.md

```diff
@@ -0,0 +1,370 @@
+# Apollo Android design discussion
+
+These are some preliminary ideas for mapping GraphQL query results to Java, based on what we currently do for Swift in Apollo iOS.
+
+## Query classes
+
+I would expect us to generate classes for queries (and other operations). Instances of these classes represent a particular query, fully configured with the required variables.
+
+These query objects can be used multiple times to parse a response or to load results from the cache. They can also be stored by a query manager that watches the store and reloads results when the contents change for example.
+
+We'll have to figure out what makes sense for Java, but one option seems to be to specify a generic type parameter on the `GraphQLQuery` interface to refer to the specific `Data` type that represents results for this query. (I've used `Data` here because this corresponds to the `data` field in a GraphQL response.)
+
+So something like this:
+
+```graphql
+query HeroAndFriendsNames($episode: Episode) {
+  hero(episode: $episode) {
+    name
+    friends {
+      name
+    }
+  }
+}
+```
+
+```java
+interface GraphQLQuery<D> {
+  D parseResponse(...);
+  // etc.
+}
+
+enum Episode {
+    NEWHOPE,
+    EMPIRE,
+    JEDI
+}
+
+class HeroAndFriendsNameQuery implements GraphQLQuery<HeroAndFriendsNameQuery.Data> {
```

Comment:
Well, I still think we want to differentiate between a fully configured query (with values for variables) and the *result* of a query. The idea is that these query instances can be used to parse a result from a server response, read from the cache, etc. So the first result could come from the server for instance, but then after another query changes the same objects, the first query can be reloaded from the (updated) cache.